### PR TITLE
[LLP] In-memory fan-out buffer size 0 allowed [DPP-1125]

### PR DESCRIPTION
**Pull request by tudor-da**
This PR allows using 0-sized in-memory fan-out buffers and other test adjustments
 * Allow 0-sized buffer
 * Introduces a metric for gauging the buffer size
 * Adapt Sandbox-on-X conformance tests

### CODE REVIEW

File: ledger/sandbox-on-x/BUILD.bazel [hunk 1]
```diff
@@ -518,7 +520,7 @@ conformance_test(
 )
 
 conformance_test(
-    name = "conformance-test-in-memory-fan-out",
```
Comment (tudor-da) [inline ledger/sandbox-on-x/BUILD.bazel:521]: NOTE: in-memory fan-out is enabled by default

Comment (skisel-da) [reply ledger/sandbox-on-x/BUILD.bazel:521]: maybe then it makes sense to remove `enableInMemoryFanOutForLedgerApi` from `IndexServiceConfig`?

Comment (tudor-da) [reply ledger/sandbox-on-x/BUILD.bazel:521]: WIll do in a later stage so we have a transition period for clients.

File: ledger/participant-integration-api/src/main/scala/platform/store/cache/EventsBuffer.scala [hunk 2]
```diff
@@ -60,19 +84,16 @@ class EventsBuffer[ENTRY](
             throw UnorderedException(lastOffset, offset)
           case _ =>
         }
-        if (_bufferLog.size.toLong == maxBufferSize) {
+        val currentBufferSize = _bufferLog.size
+        bufferSizeHistogram.update(currentBufferSize)
```
Comment (tudor-da) [inline ledger/participant-integration-api/src/main/scala/platform/store/cache/EventsBuffer.scala:88]: Metric for buffer size introduced

File: ledger/participant-integration-api/src/main/scala/platform/store/cache/EventsBuffer.scala [hunk 3]
```diff
@@ -118,28 +137,67 @@ class EventsBuffer[ENTRY](
...
-private[platform] object BufferSlice {
+private[cache] class EmptyEventsBuffer[E] extends EventsBuffer[E] {
+  override def push(offset: Offset, entry: E): Unit = ()
+
+  override def slice[FILTER_RESULT](
+      startExclusive: Offset,
+      endInclusive: Offset,
+      filter: E => Option[FILTER_RESULT],
+  ): BufferSlice[(Offset, FILTER_RESULT)] =
+    BufferSlice.LastBufferChunkSuffix(endInclusive, Vector.empty)
```
Comment (simonmaxen-da) [inline ledger/participant-integration-api/src/main/scala/platform/store/cache/EventsBuffer.scala:151]: Could name params here as an Exclusive param is being (correctly) passed an Inclusive variable.

**Review by simonmaxen-da — APPROVED** [review]

File: ledger/participant-integration-api/src/main/scala/platform/store/cache/EventsBuffer.scala (hunk 3 above)
Comment (tudor-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/cache/EventsBuffer.scala:152]: Right. Done 👍

**Comment by tudor-da** [comment]
@simonmaxen-da Since this is not controversial and it is blocking development on other fronts, will merge as is and please let me know if you have further comments.

**Pull request merged**